byteps/server/server.cc

```diff
@@ -80,26 +84,61 @@ void BytePSServerEngineThread(int i) {
     CHECK(msg.dst);
     CHECK(msg.src);
 
+    auto iter = compressor_map_.find(msg.key);
+    if (iter != compressor_map_.end()) {
+      // compress
+      if (msg.ops == ALL_RECV) {
+        common::compressor::tensor_t grad(reinterpret_cast<char*>(msg.src),
+                                          msg.len, msg.type.dtype);
+        auto compressed = iter->second->Compress(grad);
+        // 1. compress
+        auto& updates = update_buf_[msg.key];
```

Comment:
I suspect that we need a global mutex for update_buf_ now

Reply:
No.  A compression job is submitted and executed only when all data are received (ALL_RECV). So it is only done once in terms of a certain key during a pushpull operation. There is no race condition.

Reply:
But you remind me that there may be such a potential bug. I check the code again and suspect that SUM_RECV may suffer from this. There may be more than one threads writing to the same `store_[key]` simultaneously.  
https://github.com/bytedance/byteps/blob/cf020c97fc718ca209cbadbfac4cffa5e49d7d21/byteps/server/server.cc#L130-L152

The master branch may also suffer from this problem, with threads simultaneously writing to `updates.merged.tensor`.
https://github.com/bytedance/byteps/blob/cf020c97fc718ca209cbadbfac4cffa5e49d7d21/byteps/server/server.cc#L254